src/analysis/inner_product_space/l2_space.lean

```diff
@@ -95,4 +142,284 @@ lemma inner_eq_tsum (f g : lp G 2) : ⟪f, g⟫ = ∑' i, ⟪f i, g i⟫ := rfl
+  has_sum (λ i, V i (w i)) ((hV.linear_isometry_equiv hV').symm w) :=
+by simp [orthogonal_family.linear_isometry_equiv, orthogonal_family.has_sum_linear_isometry]
+
+@[simp] protected lemma linear_isometry_equiv_symm_apply_single [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) {i : ι} (x : G i) :
+  (hV.linear_isometry_equiv hV').symm (lp.single 2 i x) = V i x :=
+by simp [orthogonal_family.linear_isometry_equiv, orthogonal_family.linear_isometry_apply_single]
+
+@[simp] protected lemma linear_isometry_equiv_symm_apply_dfinsupp_sum_single
+  [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) (W₀ : Π₀ (i : ι), G i) :
+  (hV.linear_isometry_equiv hV').symm (W₀.sum (lp.single 2)) = (W₀.sum (λ i, V i)) :=
+by simp [orthogonal_family.linear_isometry_equiv,
+  orthogonal_family.linear_isometry_apply_dfinsupp_sum_single]
+
+@[simp] protected lemma linear_isometry_equiv_apply_dfinsupp_sum_single
+  [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) (W₀ : Π₀ (i : ι), G i) :
+  (hV.linear_isometry_equiv hV' (W₀.sum (λ i, V i)) : Π i, G i) = W₀ :=
+begin
+  rw ← hV.linear_isometry_equiv_symm_apply_dfinsupp_sum_single hV',
+  rw linear_isometry_equiv.apply_symm_apply,
+  ext i,
+  simp [dfinsupp.sum, lp.single_apply] {contextual := tt},
+end
+
+end orthogonal_family
+
+section
+variables (ι) (𝕜) (E)
+
+/-- A Hilbert basis on `ι` for an inner product space `E` is an identification of `E` with the `lp`
+space `ℓ²(ι, 𝕜)`. -/
+-- `nolint` because (of course) whether `E` has a Hilbert basis indexed by `ι` depends on the
+-- cardinality of `ι`.
+@[nolint has_inhabited_instance] structure hilbert_basis := of_repr :: (repr : E ≃ₗᵢ[𝕜] ℓ²(ι, 𝕜))
+
+end
+
+namespace hilbert_basis
+
+/-- `b i` is the `i`th basis vector. -/
+instance : has_coe_to_fun (hilbert_basis ι 𝕜 E) (λ _, ι → E) :=
+{ coe := λ b i, b.repr.symm (lp.single 2 i (1:𝕜)) }
+
+@[simp] protected lemma repr_symm_single (b : hilbert_basis ι 𝕜 E) (i : ι) :
+  b.repr.symm (lp.single 2 i (1:𝕜)) = b i :=
+rfl
+
+@[simp] protected lemma repr_self (b : hilbert_basis ι 𝕜 E) (i : ι) :
+  b.repr (b i) = lp.single 2 i (1:𝕜) :=
+by rw [← b.repr_symm_single, linear_isometry_equiv.apply_symm_apply]
+
+protected lemma repr_apply_apply (b : hilbert_basis ι 𝕜 E) (v : E) (i : ι) :
+  b.repr v i = ⟪b i, v⟫ :=
+begin
+  rw [← b.repr.inner_map_map (b i) v, b.repr_self, lp.inner_single_left],
+  simp,
+end
+
+@[simp] protected lemma orthonormal (b : hilbert_basis ι 𝕜 E) : orthonormal 𝕜 b :=
+begin
+  rw orthonormal_iff_ite,
+  intros i j,
+  rw [← b.repr.inner_map_map (b i) (b j), b.repr_self, b.repr_self, lp.inner_single_left,
+    lp.single_apply],
+  simp,
+end
+
+protected lemma has_sum_repr_symm (b : hilbert_basis ι 𝕜 E) (f : ℓ²(ι, 𝕜)) :
+  has_sum (λ i, f i • b i) (b.repr.symm f) :=
+begin
+  suffices H : (λ (i : ι), f i • b i) =
```

Comment:
I think a way to help the kernel here would be to go fully classical, i.e., have a definition of `lp.single` without the decidability instance parameter (even better, an irreducible definition). This can wait for another PR, of course.

Reply:
That hadn't occurred to me, thank you for the suggestion!  I'll try it in a later PR.